File: src/openapi_parser.py
#!/usr/bin/env python3
"""
OpenAPI/Swagger YAML parser for generating HTTP request files.
"""

import yaml
from typing import Dict, List, Optional, Any
from urllib.parse import urljoin


class OpenAPIParser:
    """Parse OpenAPI/Swagger YAML files and extract endpoint information."""
    
    def __init__(self, yaml_file: str):
        """
        Initialize OpenAPI parser.
        
        Args:
            yaml_file: Path to OpenAPI/Swagger YAML file
        """
        with open(yaml_file, 'r') as f:
            self.spec = yaml.safe_load(f)
        
        self.base_url = self._build_base_url()
    
    def _build_base_url(self) -> str:
        """Build base URL from spec."""
        scheme = self.spec.get('schemes', ['https'])[0]
        host = self.spec.get('host', '')
        base_path = self.spec.get('basePath', '')
        
        return f"{scheme}://{host}{base_path}"
    
    def _get_security_headers(self, operation: Dict[str, Any]) -> Dict[str, str]:
        """Extract security headers from operation."""
        headers = {}
        
        # Check operation-level security
        security = operation.get('security', [])
        if not security:
            # Check global security
            security = self.spec.get('security', [])
        
        security_defs = self.spec.get('securityDefinitions', {})
        
        for sec in security:
            for sec_name, sec_value in sec.items():
                if sec_name in security_defs:
                    sec_def = security_defs[sec_name]
                    if sec_def.get('type') == 'apiKey' and sec_def.get('in') == 'header':
                        header_name = sec_def.get('name', '')
                        # Use placeholder for token
                        headers[header_name] = f"Bearer <your-token>" if 'Bearer' in sec_def.get('description', '') else "<your-token>"
        
        return headers
    
    def _get_request_body(self, operation: Dict[str, Any]) -> Optional[str]:
        """Extract example request body from operation."""
        import json
        
        parameters = operation.get('parameters', [])
        
        for param in parameters:
            if param.get('in') == 'body':
                schema = param.get('schema', {})
                # Try to get example from schema
                if 'example' in schema:
                    return json.dumps(schema['example'], indent=2)
                # Try to get example from definition reference
                ref = schema.get('$ref', '')
                if ref and ref.startswith('#/definitions/'):
                    def_name = ref.split('/')[-1]
                    definition = self.spec.get('definitions', {}).get(def_name, {})
                    if 'example' in definition:
                        return json.dumps(definition['example'], indent=2)
                    # Generate example from properties with better defaults
                    if 'properties' in definition:
                        example = {}
                        required = definition.get('required', [])
                        
                        for prop_name, prop_schema in definition.get('properties', {}).items():
                            # Use example if provided
                            if 'example' in prop_schema:
                                example[prop_name] = prop_schema['example']
                                continue
                            
                            prop_type = prop_schema.get('type', 'string')
                            if prop_type == 'string':
                                # Check for format
                                if prop_schema.get('format') == 'email':
                                    example[prop_name] = "user@example.com"
                                elif prop_schema.get('format') == 'date-time':
                                    example[prop_name] = "2024-12-31T23:59:59Z"
                                else:
                                    example[prop_name] = f"<{prop_name}>"
                            elif prop_type == 'integer':
                                example[prop_name] = prop_schema.get('default', 0)
                            elif prop_type == 'boolean':
                                example[prop_name] = prop_schema.get('default', False)
                            elif prop_type == 'array':
                                items = prop_schema.get('items', {})
                                if items.get('$ref'):
                                    # Reference to another definition
                                    item_def_name = items['$ref'].split('/')[-1]
                                    item_def = self.spec.get('definitions', {}).get(item_def_name, {})
                                    if 'properties' in item_def:
                                        item_example = {}
                                        for item_prop, item_prop_schema in item_def['properties'].items():
                                            if 'example' in item_prop_schema:
                                                item_example[item_prop] = item_prop_schema['example']
                                            else:
                                                item_example[item_prop] = f"<{item_prop}>"
                                        example[prop_name] = [item_example]
                                    else:
                                        example[prop_name] = []
                                else:
                                    example[prop_name] = []
                            elif prop_type == 'object':
                                example[prop_name] = {}
                            
                            # Only include required fields or fields with defaults
                            if prop_name not in required and prop_name not in example:
                                continue
                        
                        return json.dumps(example, indent=2)
                # Handle inline schema (not a reference)
                elif 'type' in schema and schema.get('type') == 'object':
                    if 'properties' in schema:
                        example = {}
                        required = schema.get('required', [])
                        
                        for prop_name, prop_schema in schema.get('properties', {}).items():
                            # Use example if provided
                            if 'example' in prop_schema:
                                example[prop_name] = prop_schema['example']
                                continue
                            
                            prop_type = prop_schema.get('type', 'string')
                            if prop_type == 'string':
                                if prop_schema.get('format') == 'email':
                                    example[prop_name] = "user@example.com"
                                elif prop_schema.get('format') == 'date-time':
                                    example[prop_name] = "2024-12-31T23:59:59Z"
                                else:
                                    example[prop_name] = f"<{prop_name}>"
                            elif prop_type == 'integer':
                                example[prop_name] = prop_schema.get('default', 0)
                            elif prop_type == 'boolean':
                                example[prop_name] = prop_schema.get('default', False)
                            elif prop_type == 'array':
                                example[prop_name] = []
                            elif prop_type == 'object':
                                example[prop_name] = {}
                            
                            # Include required fields
                            if prop_name in required or prop_name in example:
                                pass  # Already added
                        
                        return json.dumps(example, indent=2)
                        example = {}
                        required = definition.get('required', [])
                        
                        for prop_name, prop_schema in definition.get('properties', {}).items():
                            # Use example if provided
                            if 'example' in prop_schema:
                                example[prop_name] = prop_schema['example']
                                continue
                            
                            prop_type = prop_schema.get('type', 'string')
                            if prop_type == 'string':
                                # Check for format
                                if prop_schema.get('format') == 'email':
                                    example[prop_name] = "user@example.com"
                                elif prop_schema.get('format') == 'date-time':
                                    example[prop_name] = "2024-12-31T23:59:59Z"
                                else:
                                    example[prop_name] = f"<{prop_name}>"
                            elif prop_type == 'integer':
                                example[prop_name] = prop_schema.get('default', 0)
                            elif prop_type == 'boolean':
                                example[prop_name] = prop_schema.get('default', False)
                            elif prop_type == 'array':
                                items = prop_schema.get('items', {})
                                if items.get('$ref'):
                                    # Reference to another definition
                                    item_def_name = items['$ref'].split('/')[-1]
                                    item_def = self.spec.get('definitions', {}).get(item_def_name, {})
                                    if 'properties' in item_def:
                                        item_example = {}
                                        for item_prop, item_prop_schema in item_def['properties'].items():
                                            if 'example' in item_prop_schema:
                                                item_example[item_prop] = item_prop_schema['example']
                                            else:
                                                item_example[item_prop] = f"<{item_prop}>"
                                        example[prop_name] = [item_example]
                                    else:
                                        example[prop_name] = []
                                else:
                                    example[prop_name] = []
                            elif prop_type == 'object':
                                example[prop_name] = {}
                            
                            # Only include required fields or fields with defaults
                            if prop_name not in required and prop_name not in example:
                                continue
                        
                        return json.dumps(example, indent=2)
        
        return None
    
    def _build_url(self, path: str, operation: Dict[str, Any]) -> str:
        """Build full URL with path parameters."""
        # In OpenAPI 2.0, paths are appended to basePath
        # basePath already includes leading /, paths also start with /
        # So we need to join them properly
        if path.startswith('/'):
            # Remove leading slash and append to base_url
            url = self.base_url.rstrip('/') + path
        else:
            url = urljoin(self.base_url.rstrip('/') + '/', path)
        
        # Replace path parameters with examples
        parameters = operation.get('parameters', [])
        for param in parameters:
            if param.get('in') == 'path':
                param_name = param.get('name', '')
                example = param.get('example', f"<{param_name}>")
                url = url.replace(f"{{{param_name}}}", str(example))
        
        # Add query parameters
        query_params = []
        for param in parameters:
            if param.get('in') == 'query':
                param_name = param.get('name', '')
                example = param.get('example', param.get('default', f"<{param_name}>"))
                query_params.append(f"{param_name}={example}")
        
        if query_params:
            url += "?" + "&".join(query_params)
        
        return url
    
    def extract_endpoints(self) -> List[Dict[str, Any]]:
        """Extract all endpoints from the OpenAPI spec."""
        endpoints = []
        paths = self.spec.get('paths', {})
        
        for path, path_item in paths.items():
            # Get common parameters for this path
            common_params = path_item.get('parameters', [])
            
            for method, operation in path_item.items():
                if method not in ['get', 'post', 'put', 'patch', 'delete', 'head', 'options']:
                    continue
                
                # Merge common parameters with operation parameters
                op_params = operation.get('parameters', [])
                all_params = common_params + op_params
                operation['parameters'] = all_params
                
                # Build URL
                url = self._build_url(path, operation)
                
                # Get headers
                headers = self._get_security_headers(operation)
                
                # Add Content-Type if there's a body
                request_body = self._get_request_body(operation)
                if request_body:
                    headers['Content-Type'] = 'application/json'
                
                # Get summary/description
                summary = operation.get('summary', '')
                description = operation.get('description', '')
                operation_id = operation.get('operationId', '')
                
                endpoints.append({
                    'method': method.upper(),
                    'url': url,
                    'headers': headers,
                    'body': request_body,
                    'summary': summary,
                    'description': description,
                    'operation_id': operation_id,
                    'path': path,
                    'tags': operation.get('tags', [])
                })
        
        return endpoints


def generate_http_file_from_openapi(yaml_file: str, output_file: str = None) -> str:
    """
    Generate HTTP request file from OpenAPI/Swagger YAML.
    
    Args:
        yaml_file: Path to OpenAPI YAML file
        output_file: Optional output file path
    
    Returns:
        Generated HTTP file content
    """
    parser = OpenAPIParser(yaml_file)
    endpoints = parser.extract_endpoints()
    
    # Group by tags
    from collections import defaultdict
    by_tag = defaultdict(list)
    for endpoint in endpoints:
        tags = endpoint.get('tags', [''])
        tag = tags[0] if tags else 'Other'
        by_tag[tag].append(endpoint)
    
    lines = []
    lines.append("### Generated from OpenAPI/Swagger specification")
    lines.append(f"### Base URL: {parser.base_url}")
    lines.append("")
    
    # Generate requests grouped by tag
    for tag, tag_endpoints in sorted(by_tag.items()):
        lines.append(f"### {tag}")
        lines.append("")
        
        for endpoint in tag_endpoints:
            # Add comment with summary
            if endpoint['summary']:
                lines.append(f"### {endpoint['summary']}")
            if endpoint['description']:
                # Take first line of description
                desc_line = endpoint['description'].split('\n')[0].strip()
                if desc_line and desc_line != endpoint['summary']:
                    lines.append(f"### {desc_line}")
            
            # Generate request
            request_lines = [f"{endpoint['method']} {endpoint['url']}"]
            
            # Add headers
            for key, value in endpoint['headers'].items():
                request_lines.append(f"{key}: {value}")
            
            # Add body
            if endpoint['body']:
                request_lines.append("")
                request_lines.append(endpoint['body'])
            
            lines.extend(request_lines)
            lines.append("")
            lines.append("")
    
    content = "\n".join(lines)
    
    if output_file:
        with open(output_file, 'w') as f:
            f.write(content)
    
    return content

File: ports/c-sharp/OpenApiParser.cs
using System.Text.Json;
using YamlDotNet.Serialization;

namespace HttpFileMaker;

/// <summary>
/// Parse OpenAPI/Swagger YAML files and extract endpoint information.
/// </summary>
public class OpenApiParser
{
    private readonly Dictionary<string, object?> _spec;
    private readonly string _baseUrl;

    public OpenApiParser(string yamlFile)
    {
        var deserializer = new DeserializerBuilder()
            .Build();

        var yamlContent = File.ReadAllText(yamlFile);
        _spec = deserializer.Deserialize<Dictionary<string, object?>>(yamlContent) 
            ?? throw new InvalidOperationException("Failed to parse YAML file");
        
        _baseUrl = BuildBaseUrl();
    }

    public string BaseUrl => _baseUrl;

    private string BuildBaseUrl()
    {
        var schemes = GetValue<List<object>>(_spec, "schemes") ?? new List<object> { "https" };
        var scheme = schemes[0].ToString() ?? "https";
        var host = GetValue<string>(_spec, "host") ?? "";
        var basePath = GetValue<string>(_spec, "basePath") ?? "";

        return $"{scheme}://{host}{basePath}";
    }

    private Dictionary<string, string> GetSecurityHeaders(Dictionary<string, object?> operation)
    {
        var headers = new Dictionary<string, string>();

        // Check operation-level security
        var security = GetValue<List<object>>(operation, "security");
        if (security == null || security.Count == 0)
        {
            // Check global security
            security = GetValue<List<object>>(_spec, "security");
        }

        var securityDefs = GetValue<Dictionary<string, object?>>(_spec, "securityDefinitions") 
            ?? new Dictionary<string, object?>();

        if (security != null)
        {
            foreach (var secObj in security)
            {
                if (secObj is Dictionary<object, object?> sec)
                {
                    foreach (var (secName, _) in sec)
                    {
                        var secNameStr = secName.ToString() ?? "";
                        if (securityDefs.TryGetValue(secNameStr, out var secDefObj) && secDefObj is Dictionary<string, object?> secDef)
                        {
                            var type = GetValue<string>(secDef, "type");
                            var location = GetValue<string>(secDef, "in");
                            if (type == "apiKey" && location == "header")
                            {
                                var headerName = GetValue<string>(secDef, "name") ?? "";
                                var description = GetValue<string>(secDef, "description") ?? "";
                                var tokenValue = description.Contains("Bearer", StringComparison.OrdinalIgnoreCase) 
                                    ? "Bearer <your-token>" 
                                    : "<your-token>";
                                headers[headerName] = tokenValue;
                            }
                        }
                    }
                }
            }
        }

        return headers;
    }

    private string? GetRequestBody(Dictionary<string, object?> operation)
    {
        var parameters = GetValue<List<object>>(operation, "parameters") ?? new List<object>();

        foreach (var paramObj in parameters)
        {
            if (paramObj is Dictionary<string, object?> param)
            {
                var paramIn = GetValue<string>(param, "in");
                if (paramIn == "body")
                {
                    var schema = GetValue<Dictionary<string, object?>>(param, "schema") 
                        ?? new Dictionary<string, object?>();

                    // Try to get example from schema
                    if (schema.TryGetValue("example", out var example))
                    {
                        return JsonSerializer.Serialize(example, new JsonSerializerOptions { WriteIndented = true });
                    }

                    // Try to get example from definition reference
                    var refValue = GetValue<string>(schema, "$ref");
                    if (!string.IsNullOrEmpty(refValue) && refValue.StartsWith("#/definitions/"))
                    {
                        var defName = refValue.Split('/').Last();
                        var definitions = GetValue<Dictionary<string, object?>>(_spec, "definitions") 
                            ?? new Dictionary<string, object?>();
                        
                        if (definitions.TryGetValue(defName, out var defObj) && defObj is Dictionary<string, object?> definition)
                        {
                            if (definition.TryGetValue("example", out var defExample))
                            {
                                return JsonSerializer.Serialize(defExample, new JsonSerializerOptions { WriteIndented = true });
                            }

                            // Generate example from properties
                            if (definition.TryGetValue("properties", out var defPropsObj) && defPropsObj is Dictionary<string, object?> defProperties)
                            {
                                var defExample = GenerateExampleFromProperties(defProperties, definition);
                                return JsonSerializer.Serialize(defExample, new JsonSerializerOptions { WriteIndented = true });
                            }
                        }
                    }
                    // Handle inline schema
                    else if (GetValue<string>(schema, "type") == "object")
                    {
                        if (schema.TryGetValue("properties", out var inlinePropsObj) && inlinePropsObj is Dictionary<string, object?> inlineProperties)
                        {
                            var inlineExample = GenerateExampleFromProperties(inlineProperties, schema);
                            return JsonSerializer.Serialize(inlineExample, new JsonSerializerOptions { WriteIndented = true });
                        }
                    }
                }
            }
        }

        return null;
    }

    private Dictionary<string, object> GenerateExampleFromProperties(
        Dictionary<string, object?> properties, 
        Dictionary<string, object?> definition)
    {
        var example = new Dictionary<string, object>();
        var required = GetValue<List<object>>(definition, "required") ?? new List<object>();
        var requiredStrings = required.Select(r => r.ToString() ?? "").ToList();

        foreach (var (propName, propSchemaObj) in properties)
        {
            if (propSchemaObj is not Dictionary<string, object?> propSchema)
                continue;

            // Use example if provided
            if (propSchema.TryGetValue("example", out var propExample))
            {
                example[propName] = propExample ?? "";
                continue;
            }

            var propType = GetValue<string>(propSchema, "type") ?? "string";

            switch (propType)
            {
                case "string":
                    var format = GetValue<string>(propSchema, "format");
                    example[propName] = format switch
                    {
                        "email" => "user@example.com",
                        "date-time" => "2024-12-31T23:59:59Z",
                        _ => $"<{propName}>"
                    };
                    break;
                case "integer":
                    example[propName] = GetValue<int?>(propSchema, "default") ?? 0;
                    break;
                case "boolean":
                    example[propName] = GetValue<bool?>(propSchema, "default") ?? false;
                    break;
                case "array":
                    var items = GetValue<Dictionary<string, object?>>(propSchema, "items");
                    if (items != null && items.TryGetValue("$ref", out var itemRef) && itemRef is string itemRefStr)
                    {
                        var itemDefName = itemRefStr.Split('/').Last();
                        var definitions = GetValue<Dictionary<string, object?>>(_spec, "definitions") ?? new Dictionary<string, object?>();
                        if (definitions.TryGetValue(itemDefName, out var itemDefObj) && itemDefObj is Dictionary<string, object?> itemDef)
                        {
                            if (itemDef.TryGetValue("properties", out var itemPropsObj) && itemPropsObj is Dictionary<string, object?> itemProps)
                            {
                                var itemExample = GenerateExampleFromProperties(itemProps, itemDef);
                                example[propName] = new[] { itemExample };
                            }
                            else
                            {
                                example[propName] = Array.Empty<object>();
                            }
                        }
                        else
                        {
                            example[propName] = Array.Empty<object>();
                        }
                    }
                    else
                    {
                        example[propName] = Array.Empty<object>();
                    }
                    break;
                case "object":
                    example[propName] = new Dictionary<string, object>();
                    break;
            }

            // Only include required fields or fields we've added
            if (!requiredStrings.Contains(propName) && !example.ContainsKey(propName))
                continue;
        }

        return example;
    }

    private string BuildUrl(string path, Dictionary<string, object?> operation)
    {
        // In OpenAPI 2.0, paths are appended to basePath
        var url = path.StartsWith('/') 
            ? _baseUrl.TrimEnd('/') + path 
            : _baseUrl.TrimEnd('/') + "/" + path;

        // Replace path parameters with examples
        var parameters = GetValue<List<object>>(operation, "parameters") ?? new List<object>();
        foreach (var paramObj in parameters)
        {
            if (paramObj is Dictionary<string, object?> param)
            {
                var paramIn = GetValue<string>(param, "in");
                if (paramIn == "path")
                {
                    var paramName = GetValue<string>(param, "name") ?? "";
                    var example = GetValue<string>(param, "example") ?? $"<{paramName}>";
                    url = url.Replace($"{{{paramName}}}", example);
                }
            }
        }

        // Add query parameters
        var queryParams = new List<string>();
        foreach (var paramObj in parameters)
        {
            if (paramObj is Dictionary<string, object?> param)
            {
                var paramIn = GetValue<string>(param, "in");
                if (paramIn == "query")
                {
                    var paramName = GetValue<string>(param, "name") ?? "";
                    var example = GetValue<string>(param, "example") 
                        ?? GetValue<string>(param, "default") 
                        ?? $"<{paramName}>";
                    queryParams.Add($"{paramName}={example}");
                }
            }
        }

        if (queryParams.Count > 0)
        {
            url += "?" + string.Join("&", queryParams);
        }

        return url;
    }

    public List<Endpoint> ExtractEndpoints()
    {
        var endpoints = new List<Endpoint>();
        var paths = GetValue<Dictionary<string, object?>>(_spec, "paths") 
            ?? new Dictionary<string, object?>();

        foreach (var (path, pathItemObj) in paths)
        {
            if (pathItemObj is not Dictionary<string, object?> pathItem)
                continue;

            // Get common parameters for this path
            var commonParams = GetValue<List<object>>(pathItem, "parameters") ?? new List<object>();

            foreach (var (method, operationObj) in pathItem)
            {
                if (!new[] { "get", "post", "put", "patch", "delete", "head", "options" }.Contains(method.ToLowerInvariant()))
                    continue;

                if (operationObj is not Dictionary<string, object?> operation)
                    continue;

                // Merge common parameters with operation parameters
                var opParams = GetValue<List<object>>(operation, "parameters") ?? new List<object>();
                var allParams = new List<object>(commonParams);
                allParams.AddRange(opParams);
                operation["parameters"] = allParams;

                // Build URL
                var url = BuildUrl(path, operation);

                // Get headers
                var headers = GetSecurityHeaders(operation);

                // Add Content-Type if there's a body
                var requestBody = GetRequestBody(operation);
                if (!string.IsNullOrEmpty(requestBody))
                {
                    headers["Content-Type"] = "application/json";
                }

                // Get summary/description
                var summary = GetValue<string>(operation, "summary") ?? "";
                var description = GetValue<string>(operation, "description") ?? "";
                var operationId = GetValue<string>(operation, "operationId") ?? "";
                var tags = GetValue<List<object>>(operation, "tags") ?? new List<object>();

                endpoints.Add(new Endpoint
                {
                    Method = method.ToUpperInvariant(),
                    Url = url,
                    Headers = headers,
                    Body = requestBody,
                    Summary = summary,
                    Description = description,
                    OperationId = operationId,
                    Path = path,
                    Tags = tags.Select(t => t.ToString() ?? "").ToList()
                });
            }
        }

        return endpoints;
    }

    private static T? GetValue<T>(Dictionary<string, object?> dict, string key)
    {
        if (!dict.TryGetValue(key, out var value) || value == null)
            return default;

        if (value is T directValue)
            return directValue;

        // Handle nested dictionaries
        if (typeof(T) == typeof(Dictionary<string, object?>) && value is Dictionary<object, object?> nestedDict)
        {
            var result = new Dictionary<string, object?>();
            foreach (var (k, v) in nestedDict)
            {
                result[k.ToString() ?? ""] = v;
            }
            return (T)(object)result;
        }

        // Handle lists
        if (typeof(T).IsGenericType && 
            typeof(T).GetGenericTypeDefinition() == typeof(List<>))
        {
            var elementType = typeof(T).GetGenericArguments()[0];
            var listType = typeof(List<>).MakeGenericType(elementType);
            var result = Activator.CreateInstance(listType) as System.Collections.IList;
            if (result != null && value is System.Collections.IEnumerable enumerable)
            {
                foreach (var item in enumerable)
                {
                    result.Add(item);
                }
                return (T)result!;
            }
        }

        // Try to convert
        try
        {
            return (T)Convert.ChangeType(value, typeof(T));
        }
        catch
        {
            return default;
        }
    }
}

public class Endpoint
{
    public string Method { get; set; } = "";
    public string Url { get; set; } = "";
    public Dictionary<string, string> Headers { get; set; } = new();
    public string? Body { get; set; }
    public string Summary { get; set; } = "";
    public string Description { get; set; } = "";
    public string OperationId { get; set; } = "";
    public string Path { get; set; } = "";
    public List<string> Tags { get; set; } = new();
}

